libraries/platforms/dromajo-vcs/bsg_manycore_simulator.cpp

```diff
@@ -0,0 +1,619 @@
+ * @params[in] hierarchy - A C++ string that holds the manycore hierarchy
+ * @returns success if initialized correctly
+ */
+int SimulationWrapper::dpi_init() {
+  advance_time();
+
+  int err;
+
+  err = dpi_fifo_drain(HB_MC_FIFO_RX_REQ);
+  if (err != HB_MC_SUCCESS) {
+    dpi_cleanup();
+    return err;
+  }
+
+  err = dpi_fifo_drain(HB_MC_FIFO_RX_RSP);
+  if (err != HB_MC_SUCCESS) {
+    dpi_cleanup();
+    return err;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_init
+ * Initializes dromajo with the correct command-line arguments
+ * @returns a pointer to a dromajo_cosim_state_t object
+ */
+int SimulationWrapper::dromajo_init() {
+  char dromajo_str[50];
+  char host_str[50];
+  char manycore_str[50];
+  char prog_str[50];
+
+  sprintf(dromajo_str, "dromajo");
+  sprintf(host_str, "--host");
+  sprintf(manycore_str, "--manycore");
+  sprintf(prog_str, "main.elf");
+
+  char* argv[] = {dromajo_str, host_str, manycore_str, prog_str};
+  dromajo = dromajo_cosim_init(4, argv);
+  if (!dromajo) {
+    bsg_pr_err("Failed to initialize Dromajo!\n");
+    return HB_MC_FAIL;
+  }
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_step
+ * Executes 1 instruction in Dromajo
+ * @returns true if execution is incomplete and/or without errors
+ */
+bool SimulationWrapper::dromajo_step() {
+  // Execute dromajo with verbose mode on
+  int err = dromajo_cosim_step(dromajo, 0, 0, 0, 0, 0, false, false);
+  return (err != 0) ? false : true;
+}
+
+/*
+ * dromajo_transmit_packet
+ * Fetches data from the Dromajo->Manycore FIFO and pushes it into the DPI FIFO
+ * to send packets to the manycore
+ * @returns success on succesful transmission
+ */
+int SimulationWrapper::dromajo_transmit_packet() {
+  hb_mc_packet_t dromajo_to_mc_packet, host_to_dromajo_packet;
+  int err;
+  __m128i *pkt;
+
+  // Check if FIFO has an element and hence ready to transmit
+  mc_fifo_type_t type = FIFO_HOST_TO_MC_REQ;
+  bool is_empty = mc_is_fifo_empty(type);
+
+  if (!is_empty) {
+    // Read the FIFO head pointer for all 32-bit FIFOs
+    dromajo_to_mc_packet.words[0] = host_to_mc_req_fifo->fifo[0].front();
+    dromajo_to_mc_packet.words[1] = host_to_mc_req_fifo->fifo[1].front();
+    dromajo_to_mc_packet.words[2] = host_to_mc_req_fifo->fifo[2].front();
+    dromajo_to_mc_packet.words[3] = host_to_mc_req_fifo->fifo[3].front();
+
+    // Intercept packets that are for the host and generate appropriate responses
+    if ((dromajo_to_mc_packet.request.x_dst == ((0 << 4) | 0)) && (dromajo_to_mc_packet.request.y_dst == ((1 << 3) | 0))) {
+      host_to_dromajo_packet.response.x_dst = dromajo_to_mc_packet.request.x_src;
+      host_to_dromajo_packet.response.y_dst = dromajo_to_mc_packet.request.y_src;
+      host_to_dromajo_packet.response.load_id = 0;
+
+      uint32_t data = 0;
+      switch (dromajo_to_mc_packet.request.op_v2) {
+        case HB_MC_PACKET_OP_REMOTE_LOAD:
+        {
+          // Fixme: Is there no struct for response opcode like in the manycore hardware
```

Comment:
No, but there should be :)

Reply:
Is this something that needs to be handled in this PR?

Reply:
No, but we should probably file an issue. It might be in my network simulator PR

Reply:
Is this what you're looking for? https://github.com/bespoke-silicon-group/bsg_replicant/blob/f29dccb246f17efcfc81948e73ff6532283f998d/libraries/bsg_manycore_request_packet.h#L106

Reply:
I was wondering why there is nothing like this struct from the manycore hardware repository? https://github.com/bespoke-silicon-group/bsg_manycore/blob/d7ab95c470bc3ade5dce2a18046339ea211f6a8c/v/bsg_manycore_pkg.v#L42-L55

I don't check for return packet types right now so there shouldn't be any issues but I was wondering why not having this struct defined somewhere is fine for the library?

Reply:
Because so far, we haven't needed one. At the moment the host only gets response packets when it performs a read, and we only treat data as integer, so we only get int_wb.

Reply:
I misread and thought you were asking about requests.